packages/iov-stream/src/concat.ts

```diff
@@ -0,0 +1,102 @@
+// tslint:disable:readonly-array
+import { Producer, Stream, Subscription } from "xstream";
+
+/**
+ * An implementation of concat that buffers all source stream events
+ *
+ * Marble diagram:
+ *
+ * ```text
+ * --1--2---3---4-|
+ * -a--b-c--d-|
+ * --------X---------Y---------Z-
+ *           concat
+ * --1--2---3---4-abcdXY-------Z-
+ * ```
+ *
+ * This is inspired by RxJS's concat as documented at http://rxmarbles.com/#concat and behaves
+ * differently than xstream's concat as discussed in https://github.com/staltz/xstream/issues/170.
+ *
+ */
+export function concat<T>(...streams: Array<Stream<T>>): Stream<T> {
+  const subscriptions = new Array<Subscription>();
+  const queues = new Array<T[]>(); // one queue per stream
+  const completedStreams = new Set<number>();
+  let activeStreamIndex = 0;
+
+  function reset(): void {
+    while (subscriptions.length > 0) {
+      const subscription = subscriptions.shift()!;
+      subscription.unsubscribe();
+    }
+
+    // tslint:disable-next-line:no-object-mutation
+    queues.length = 0;
```

Comment:
Huh? Setting the length empties the array?
Why not just reset it to a new Array (and use let above). Seems cleaner

Reply:
This is the only way I am aware of to clear an Array object. It is dirty as hell compared to a `clear()` method like you have for [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set)s or [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)s.

But in this case I don't want to create another object (by assigning a new array), which makes it harder to track which object is used in which code block.

Reply:
Fair enough